appengine/go11x/static/main.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2018 Google Inc. All rights reserved.
+// Use of this source code is governed by the Apache 2.0
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"fmt"
+	"html/template"
+	"log"
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+var (
+	indexTmpl = template.Must(
+		template.ParseFiles(filepath.Join("templates", "index.html")),
+	)
+)
+
+func main() {
+	http.Handle("/", withLogging(http.HandlerFunc(indexHandler)))
+
+	// Serve static files out of the public directory.
+	// By configuring a static handler in app.yaml, App Engine serves all the
+	// static content itself. As a result, the following two lines are in
+	// effect for development only.
+	public := http.StripPrefix("/public", http.FileServer(http.Dir("public")))
+	http.Handle("/public/", withLogging(public))
+
+	port := os.Getenv("PORT")
+	if port == "" {
+		port = "8080"
+		log.Printf("Defaulting to port %s", port)
+	}
+
+	log.Printf("Listening on port %s", port)
+	log.Fatal(http.ListenAndServe(fmt.Sprintf(":%s", port), nil))
+}
+
+// indexHandler uses a template to create an index.html.
+func indexHandler(w http.ResponseWriter, r *http.Request) {
+	if r.URL.Path != "/" {
+		http.NotFound(w, r)
+		return
+	}
+	type indexData struct {
+		Logo        string
+		Style       string
+		RequestTime string
+	}
+	data := indexData{
+		Logo:        "/public/gcp-gopher.svg",
+		Style:       "/public/style.css",
+		RequestTime: time.Now().Format(time.RFC822),
+	}
+	err := indexTmpl.Execute(w, data)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+// withLogging creates an http.Handler that logs the resulting status of a
+// request.
+func withLogging(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		resp := &statusResponse{ResponseWriter: w, statusCode: http.StatusOK}
+		next.ServeHTTP(resp, r)
+		log.Printf("[%d] %s %s", resp.statusCode, r.Method, r.URL.Path)
+	})
+}
+
+// statusResponse wraps http.ResponseWriter to record the response's status
+// code.
+type statusResponse struct {
```

Comment:
Not sure I understand this. Why doesn't the normal response code setup work?

Reply:
In the context of middleware, there's no way of telling what a subsequent handlers has done. The approach saves the status code for use after the subsequent handlers are complete. Happy to consider alternative approaches.

Reply:
Gotcha. It looks like `statusCode` isn't used, though? If the example is for serving static files, everything else should be minimal? I might be missing the context of the sample.

Reply:
The `statusCode` field is used in the call to `log.Printf` above.

@grayside proposed adding logging to the sample to help users see when the static assets were being served. But it might be too much of a distraction from the purpose of this sample. I'm happy to remove it.

Reply:
Ahh. I understand now. `next` can call `WriteHeader`. SGTM.